### PR TITLE
B 17588 enable prime to add partial weight

### PR DESCRIPTION
**Pull request by TevinAdams**
Ticket: B-17588

This PR will allow the prime to enter in a billable weight for each serviceitem. The weight will be entered as a key value pair with key: WeightBilled and value as the weight.

This change is for the following serviceItems:
DLH
DSH
DUPK
DNPK
DOFSIT
DOASIT 
DOPSIT
DDDSIT 
DDSHUT 
DOSHUT 
FSC 
DDASIT 

To test:
Submit a payment request with at least one of the service items listed above. Make sure the service Item has the params populated with a WeightBilled key value pair, where key is equals WeightBilled and value is the requested weight for the service Item. Submit the payment request.

Go to the TIO View and view the payment request.
View the serviceItems and notice the cwt weight. 

I do not mind huddling up and walking through how to test this. It is tricky.

### CODE REVIEW

**Comment by cameroncaci** [comment]
I thought if no weight was provided then the weight should be the entire shipment weight instead of 0? Is weight for payment handled differently than I'm thinking?

**Comment by TevinAdams** [comment]
> I thought if no weight was provided then the weight should be the entire shipment weight instead of 0? Is weight for payment handled differently than I'm thinking?

Correct. Changes were made to meet this requirment.

**Review by traskowskycaci — COMMENTED** [review]
Which service item types will this work for? I have tried a few, including DOPSIT and DOFSIT and got the responses of: 
{
  "detail": "the parameter Service Item Parameter Name is either invalid or cannot be passed while creating a payment request for a DOPSIT service item",
  "instance": "8c78492c-5bd1-47ac-ab53-6b9180eb0d7b",
  "title": "Validation Error",
  "invalidFields": {}
}

{
  "detail": "the parameter Service Item Parameter Name is either invalid or cannot be passed while creating a payment request for a DOFSIT service item",
  "instance": "8d34eb7b-7a65-484b-b962-29004972f51f",
  "title": "Validation Error",
  "invalidFields": {}
}  

Also tried DOASIT:
{
  "detail": "the parameter Service Item Parameter Name is either invalid or cannot be passed while creating a payment request for a DOASIT service item",
  "instance": "6a2a48f6-514e-4040-b6fa-57395626a70d",
  "title": "Validation Error",
  "invalidFields": {}
}

I am logged in as a prime user and am using the prime swagger UI. User error on my part?

**Comment by TevinAdams** [comment]
> Which service item types will this work for? I have tried a few, including DOPSIT and DOFSIT and got the responses of: { "detail": "the parameter Service Item Parameter Name is either invalid or cannot be passed while creating a payment request for a DOPSIT service item", "instance": "8c78492c-5bd1-47ac-ab53-6b9180eb0d7b", "title": "Validation Error", "invalidFields": {} }
> 
> { "detail": "the parameter Service Item Parameter Name is either invalid or cannot be passed while creating a payment request for a DOFSIT service item", "instance": "8d34eb7b-7a65-484b-b962-29004972f51f", "title": "Validation Error", "invalidFields": {} }
> 
> Also tried DOASIT: { "detail": "the parameter Service Item Parameter Name is either invalid or cannot be passed while creating a payment request for a DOASIT service item", "instance": "6a2a48f6-514e-4040-b6fa-57395626a70d", "title": "Validation Error", "invalidFields": {} }
> 
> I am logged in as a prime user and am using the prime swagger UI. User error on my part?

Did you put the shipmentid in the body?

**Comment by danieljordan-caci** [comment]
> Which service item types will this work for? I have tried a few, including DOPSIT and DOFSIT and got the responses of: { "detail": "the parameter Service Item Parameter Name is either invalid or cannot be passed while creating a payment request for a DOPSIT service item", "instance": "8c78492c-5bd1-47ac-ab53-6b9180eb0d7b", "title": "Validation Error", "invalidFields": {} }
> 
> { "detail": "the parameter Service Item Parameter Name is either invalid or cannot be passed while creating a payment request for a DOFSIT service item", "instance": "8d34eb7b-7a65-484b-b962-29004972f51f", "title": "Validation Error", "invalidFields": {} }
> 
> Also tried DOASIT: { "detail": "the parameter Service Item Parameter Name is either invalid or cannot be passed while creating a payment request for a DOASIT service item", "instance": "6a2a48f6-514e-4040-b6fa-57395626a70d", "title": "Validation Error", "invalidFields": {} }
> 
> I am logged in as a prime user and am using the prime swagger UI. User error on my part?

This looks like you were passing in invalid parameter keys for service items (`Service Item Parameter Name`) is not a valid parameter key - the appropriate keys can be found in the `service_item_param_keys` table.

**Comment by traskowskycaci** [comment]
> > Which service item types will this work for? I have tried a few, including DOPSIT and DOFSIT and got the responses of: { "detail": "the parameter Service Item Parameter Name is either invalid or cannot be passed while creating a payment request for a DOPSIT service item", "instance": "8c78492c-5bd1-47ac-ab53-6b9180eb0d7b", "title": "Validation Error", "invalidFields": {} }
> > { "detail": "the parameter Service Item Parameter Name is either invalid or cannot be passed while creating a payment request for a DOFSIT service item", "instance": "8d34eb7b-7a65-484b-b962-29004972f51f", "title": "Validation Error", "invalidFields": {} }
> > Also tried DOASIT: { "detail": "the parameter Service Item Parameter Name is either invalid or cannot be passed while creating a payment request for a DOASIT service item", "instance": "6a2a48f6-514e-4040-b6fa-57395626a70d", "title": "Validation Error", "invalidFields": {} }
> > I am logged in as a prime user and am using the prime swagger UI. User error on my part?
> 
> This looks like you were passing in invalid parameter keys for service items (`Service Item Parameter Name`) is not a valid parameter key - the appropriate keys can be found in the `service_item_param_keys` table.

that was indeed one of the issues. had some other issues with the mto shipment i was trying to use missing an address, so created a new move with a new shipment and new service item and that did the trick.

**Review by traskowskycaci — APPROVED** [review]
Hopped on a call with Tevin and he walked me through the process of creating a valid, testable move with associated MTO shipments, service items, and payment requests. Tested adding partial weight as the prime via swagger and it works great. Approving with the caveat that Tevin has more work to do on this to update the prime simulator.

**Review by danieljordan-caci — CHANGES_REQUESTED** [review]

File: migrations/app/schema/20231106173413_add_weight_column_to_payment_request.up.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE payment_requests
+ADD COLUMN requested_weight_amount integer;
+
+COMMENT ON COLUMN payment_requests.requested_weight_amount IS 'The desired weight amount for the payment request.';
```
[the remark this answers is not in the document]
Comment (danieljordan-caci) [reply migrations/app/schema/20231106173413_add_weight_column_to_payment_request.up.sql:4]: I just realized I wrote these comments but it didn't complete the review. Guess there was an additional step when submitting.

**Comment by danieljordan-caci** [comment]
Getting this error due to `NULL` values in the db in the `requested_weight_amount` column.
```
"unable to fetch records: sql: Scan error on column index 10, name \"requested_weight_amount\": converting NULL to int is unsupported"
```

Also instead of having an input for the shipment ID in the UI, can you just send the shipment ID that is showing on the page when submitted? Attaching screenshot.
Screenshot 2023-11-20 at 12 35 53 PM

**Comment by cameroncaci** [comment]
Marking back to draft as we await new changes per meetings with Daniel and Tevin

**Comment by TevinAdams** [comment]
> Getting this error due to `NULL` values in the db in the `requested_weight_amount` column.
> 
> ```
> "unable to fetch records: sql: Scan error on column index 10, name \"requested_weight_amount\": converting NULL to int is unsupported"
> ```
> 
> Also instead of having an input for the shipment ID in the UI, can you just send the shipment ID that is showing on the page when submitted? Attaching screenshot. Screenshot 2023-11-20 at 12 35 53 PM

You could but this is no longer needed.

**Review by traskowskycaci — APPROVED** [review]
Looks good. I see the billable weight (cwt) that matches my request.

image

File: pkg/handlers/primeapi/allowed_payment_service_item_params.go [hunk 2]
```diff
@@ -12,10 +12,42 @@ var (
...
 			models.ServiceItemParamNameSITPaymentRequestStart,
 			models.ServiceItemParamNameSITPaymentRequestEnd,
+			models.ServiceItemParamNameWeightBilled,
 		},
 		models.ReServiceCodeDDASIT: {
 			models.ServiceItemParamNameSITPaymentRequestStart,
 			models.ServiceItemParamNameSITPaymentRequestEnd,
+			models.ServiceItemParamNameWeightBilled,
+		},
+		models.ReServiceCodeDOFSIT: {
```
Comment (danieljordan-caci) [inline pkg/handlers/primeapi/allowed_payment_service_item_params.go:22]: Shouldn't this also include `DDFSIT`?

Here's the list I have:
```DOFSIT (Domestic origin 1st day SIT)
DOASIT (Domestic origin add'l SIT)
DOPSIT (Domestic origin SIT pickup)
DOSHUT (Domestic origin shuttle service)
DDFSIT (Domestic destination 1st day SIT)
DDASIT (Domestic destination add'l SIT)
DDDSIT (Domestic destination SIT delivery)
DDSHUT (Domestic destination shuttle service)

**Review by danieljordan-caci — CHANGES_REQUESTED** [review]

File: pkg/handlers/primeapi/allowed_payment_service_item_params.go (hunk 2 above)
Comment (danieljordan-caci) [reply pkg/handlers/primeapi/allowed_payment_service_item_params.go:22]: image
On this screenshot I am seeing 1st day SIT items: CWT, Days (Always = 1)

Comment (TevinAdams) [reply pkg/handlers/primeapi/allowed_payment_service_item_params.go:22]: I went by the list in the screenshot. DDFSIT = Domestic Destination First Day SIT isn't on the screenshot but I can add it. I had to cross reference the names in the screenshot with the codes in the DB so I might have miss taken one for another.

Comment (danieljordan-caci) [reply pkg/handlers/primeapi/allowed_payment_service_item_params.go:22]: It is - you can see it addresses both origin and destination: "Domestic SIT 1st Day & Warehouse"

Comment (TevinAdams) [reply pkg/handlers/primeapi/allowed_payment_service_item_params.go:22]: I'll add it.

Comment (TevinAdams) [reply pkg/handlers/primeapi/allowed_payment_service_item_params.go:22]: Added

**Review by danieljordan-caci — APPROVED** [review]
LGTM

**Pull request merged**